src/common/hist_util.cc

```diff
@@ -92,24 +92,66 @@ template void CopyHist(GHistRow<double> dst, const GHistRow<double> src,
  * \brief Compute Subtraction: dst = src1 - src2 in range [begin, end)
  */
 template<typename GradientSumT>
-void SubtractionHist(GHistRow<GradientSumT> dst, const GHistRow<GradientSumT> src1,
-                     const GHistRow<GradientSumT> src2,
+void SubtractionHist(GradientSumT* dst, const GradientSumT* src1,
+                     const GradientSumT* src2,
                      size_t begin, size_t end) {
-  GradientSumT* pdst = reinterpret_cast<GradientSumT*>(dst.data());
-  const GradientSumT* psrc1 = reinterpret_cast<const GradientSumT*>(src1.data());
-  const GradientSumT* psrc2 = reinterpret_cast<const GradientSumT*>(src2.data());
-
-  for (size_t i = 2 * begin; i < 2 * end; ++i) {
-    pdst[i] = psrc1[i] - psrc2[i];
+  for (size_t i = begin; i < end; ++i) {
+    dst[i] = src1[i] - src2[i];
   }
 }
-template void SubtractionHist(GHistRow<float> dst, const GHistRow<float> src1,
-                              const GHistRow<float> src2,
+template void SubtractionHist(float* dst, const float* src1,
+                              const float* src2,
                               size_t begin, size_t end);
-template void SubtractionHist(GHistRow<double> dst, const GHistRow<double> src1,
-                              const GHistRow<double> src2,
+template void SubtractionHist(double* dst, const double* src1,
+                              const double* src2,
                               size_t begin, size_t end);
 
+template<typename GradientSumT>
+void ClearHist(GradientSumT* dest_hist,
+                size_t begin, size_t end) {
+  for (size_t bin_id = begin; bin_id < end; ++bin_id) {
+    dest_hist[bin_id]  = 0;
+  }
+}
+template void ClearHist(float* dest_hist,
+                        size_t begin, size_t end);
+template void ClearHist(double* dest_hist,
+                        size_t begin, size_t end);
+
+template<typename GradientSumT>
+void ReduceHist(GradientSumT* dest_hist, GradientSumT* hist0,
+                const std::vector<std::vector<uint16_t>>& local_threads_mapping,
+                std::vector<std::vector<std::vector<GradientSumT>>>* histograms,
+                const size_t node_id,
+                const std::vector<uint16_t>& threads_id_for_node,
+                size_t begin, size_t end) {
+  for (size_t bin_id = begin; bin_id < end; ++bin_id) {
+    dest_hist[bin_id] = hist0[bin_id];
+    hist0[bin_id] = 0;
+  }
+  for (size_t tid = 1; tid < threads_id_for_node.size(); ++tid) {
+    const size_t thread_id = threads_id_for_node[tid];
+    const size_t mapped_nod_id = local_threads_mapping[thread_id][node_id];
+    GradientSumT* hist =  (*histograms)[thread_id][mapped_nod_id].data();
+    for (size_t bin_id = begin; bin_id < end; ++bin_id) {
+      dest_hist[bin_id] += hist[bin_id];
+      hist[bin_id] = 0;
+    }
+  }
+}
+template void ReduceHist(float* dest_hist, float* hist0,
```

Comment:
In general I'm ok with dropping single-precision and can even try to help with disabling it for cpu hist after all optimization are merged (because with currently implemented optimizations it works well enough even on double precision for previously affected cases).